pkg/reconcile/reconcile.go

```diff
@@ -112,11 +112,15 @@ type terminalError struct {
 	err error
 }
 
+// This function will return nil if te.err is nil.
 func (te *terminalError) Unwrap() error {
 	return te.err
 }
 
 func (te *terminalError) Error() string {
+	if te.err == nil {
+		return "nil terminal error"
+	}
```

Comment:
Is this a common pattern? We're at this point creating an error with a nil cause/inner error which might throw off the reconciliation loop.

Reply:
I only see it used here: https://github.com/kubernetes-sigs/controller-runtime/blob/7f0c6dc440f334849b25866dd9990ac1a5a777c2/pkg/internal/controller/controller.go#L319
which doesn't totally make sense because we just checked for `err==nil`

An alternate solution would be to modify the controller to create that error with a non-nil value (empty string?) and maybe add protection in the [TerminalError()](https://github.com/kubernetes-sigs/controller-runtime/blob/dbcf380a24f5048925a3981905ded66a72c2a35f/pkg/reconcile/reconcile.go#L107) function to not allow nil values.

Reply:
cc @alvaroaleman

Reply:
The nil terminal error there is only used for the `errors.Is` comparison, in that context, the `Error()` method will never be called.

It isn't really expected to construct a terminal error with a nil error when returning it. When did you run into this panicing?